src/mono/wasm/runtime/invoke-js.ts

```diff
@@ -29,80 +31,178 @@ export function mono_wasm_bind_js_function(function_name: MonoStringRef, module_
         }
-
-        for (let index = 0; index < args_count; index++) {
-            const sig = get_sig(signature, index + 2);
-            const marshaler_type = get_signature_type(sig);
-            if (marshaler_type == MarshalerType.Span) {
-                const arg_name = `arg${index}`;
-                body += `  ${arg_name}.dispose();\n`;
-            }
+        else if (args_count == 2 && !has_cleanup && res_converter) {
+            bound_fn = bind_fn_2R(closure);
+        }
+        else {
+            bound_fn = bind_fn(closure);
         }
 
-        body += "} catch (ex) {\n";
-        body += "  marshal_exception_to_cs(args, ex);\n";
-        body += "}}";
-        const factory = new Function("closure", body);
-        const bound_fn = factory(closure);
-        bound_fn[bound_js_function_symbol] = true;
-        const bound_function_handle = mono_wasm_get_js_handle(bound_fn)!;
-        setI32(function_js_handle, <any>bound_function_handle);
-    } catch (ex) {
+        (<any>bound_fn)[imported_js_function_symbol] = true;
+        const fn_handle = fn_wrapper_by_fn_handle.length;
+        fn_wrapper_by_fn_handle.push(bound_fn);
+        setI32(function_js_handle, <any>fn_handle);
+    } catch (ex: any) {
+        Module.printErr(ex.toString());
         wrap_error_root(is_exception, ex, resultRoot);
     } finally {
         resultRoot.release();
         function_name_root.release();
     }
 }
 
+function bind_fn_0V(closure: BindingClosure) {
+    const fn = closure.fn;
+    return function bound_fn_0V(args: JSMarshalerArguments) {
+        try {
+            // call user function
+            fn();
+        } catch (ex) {
+            marshal_exception_to_cs(<any>args, ex);
+        }
+    };
+}
+
+function bind_fn_1V(closure: BindingClosure) {
+    const fn = closure.fn;
+    const marshaler1 = closure.arg_marshalers[0]!;
+    return function bound_fn_1V(args: JSMarshalerArguments) {
+        try {
+            const arg1 = marshaler1(args);
+            // call user function
+            fn(arg1);
+        } catch (ex) {
+            marshal_exception_to_cs(<any>args, ex);
+        }
+    };
+}
+
+function bind_fn_1R(closure: BindingClosure) {
+    const fn = closure.fn;
+    const marshaler1 = closure.arg_marshalers[0]!;
+    const res_converter = closure.res_converter!;
+    return function bound_fn_1R(args: JSMarshalerArguments) {
+        try {
+            const arg1 = marshaler1(args);
+            // call user function
+            const js_result = fn(arg1);
+            res_converter(args, js_result);
+        } catch (ex) {
+            marshal_exception_to_cs(<any>args, ex);
+        }
+    };
+}
+
+function bind_fn_2R(closure: BindingClosure) {
+    const fn = closure.fn;
+    const marshaler1 = closure.arg_marshalers[0]!;
+    const marshaler2 = closure.arg_marshalers[1]!;
+    const res_converter = closure.res_converter!;
+    return function bound_fn_2R(args: JSMarshalerArguments) {
```

Comment:
please clear the 'closure' argument just to make sure we don't close over it (afaik JS runtimes generally won't close over it, but I know there are specific cases where in the past they would have held onto it)